### PR TITLE
Introduce minPoints option

### PR DESCRIPTION
**Pull request by zmiao**
port https://github.com/mapbox/supercluster/pull/156

### CODE REVIEW

**Review by mourner — APPROVED** [review]
👍

File: include/supercluster.hpp [hunk 1]
```diff
@@ -286,41 +288,63 @@ class Supercluster {
...
-                        options_.reduce(clusterProperties, *b.properties);
+                    if (!b.visited) {
+                        num_points += b.num_points;
                     }
                 });
 
-                if (num_points == 1) {
-                    clusters.emplace_back(weight, 1, p.id, clusterProperties);
-                } else {
+                auto clusterProperties = p.properties ? *p.properties : property_map{};
```
Comment (alexshalamov) [inline include/supercluster.hpp:304]: `clusterProperties` only mutated if `num_points >= options_.minPoints` and if we have reduce function. Could we share `p.properties` between clusters if no mutation is required?

File: include/supercluster.hpp [hunk 2]
```diff
@@ -286,41 +288,63 @@ class Supercluster {
...
+                    if (num_points > 1) {
+                        previous.tree.within(p.pos.x, p.pos.y, r, [&](const auto &neighbor_id) {
+                            assert(neighbor_id < cluster_size);
+                            auto &b = previous.clusters[neighbor_id];
+                            // filter out neighbors that are already processed
+                            if (b.visited) {
+                                return;
+                            }
+                            b.visited = true;
+                            clusters.emplace_back(b.pos, 1, b.id,
```
Comment (alexshalamov) [inline include/supercluster.hpp:344]: Probably, `b.properties` can be shared between clusters? Could we change `std::unique_ptr<property_map> properties{ nullptr }` to `std::shared_ptr<const property_map> properties{ nullptr };` and then try sharing properties between clusters if property map is not changed?

**Review by alexshalamov — APPROVED** [review]
Looks good 👍 

One suggestion, maybe for a follow-up / investigation work. It looks like we could avoid expensive unordered_map copies between clusters.

**Comment by zmiao** [comment]
@alexshalamov Thanks for your suggestion! After some initial experiments for property map sharing as you suggested in the comments, I think it worths further investigation, and more tests need to be conducted to make sure it won't break anything. So I created an issue https://github.com/mapbox/supercluster.hpp/issues/26 to track it.

**Pull request merged**